### PR TITLE
fix(serializer): empty object as array with supports cache 

### PR DESCRIPTION
**Pull request by soyuka**
reverts #4999 and re-implement it

### CODE REVIEW

File: src/Hydra/Serializer/CollectionNormalizer.php [hunk 1]
```diff
@@ -103,6 +108,23 @@ public function normalize(mixed $object, string $format = null, array $context =
...
     public function hasCacheableSupportsMethod(): bool
     {
-        return false;
+        return true;
+    }
+
+    /**
+     * Normalizes a raw collection (not API resources).
+     */
+    protected function normalizeRawCollection(iterable $object, string $format = null, array $context = [])
```
Comment (alanpoulain) [inline src/Hydra/Serializer/CollectionNormalizer.php:117]: Shouldn't we add the typehint?

File: src/Metadata/Resource/ResourceMetadataCollection.php [hunk 2]
```diff
@@ -51,7 +51,9 @@ public function getOperation(?string $operationName = null, bool $forceCollectio
 
             foreach ($metadata->getOperations() ?? [] as $name => $operation) {
                 $isCollection = $operation instanceof CollectionOperationInterface;
-                if ('' === $operationName && \in_array($operation->getMethod() ?? HttpOperation::METHOD_GET, [HttpOperation::METHOD_GET, HttpOperation::METHOD_OPTIONS, HttpOperation::METHOD_HEAD], true) && ($forceCollection ? $isCollection : !$isCollection)) {
+                $method = $operation->getMethod() ?? HttpOperation::METHOD_GET;
+                $isGetOperation = HttpOperation::METHOD_GET === $method || HttpOperation::METHOD_OPTIONS === $method || HttpOperation::METHOD_HEAD === $method;
+                if ('' === $operationName && $isGetOperation && ($forceCollection ? $isCollection : !$isCollection)) {
```
Comment (alanpoulain) [inline src/Metadata/Resource/ResourceMetadataCollection.php:56]: I'm taking this PR as an opportunity for a long time question I have: is adding an empty string as key in the `operationCache` a good idea?

Comment (soyuka) [reply src/Metadata/Resource/ResourceMetadataCollection.php:56]: I need to prefix these, but yeah the representation of `null` as string would be an empty string ?

File: src/Hydra/Serializer/CollectionNormalizer.php (hunk 1 above)
Comment (soyuka) [reply src/Hydra/Serializer/CollectionNormalizer.php]: right thanks!

File: src/Hydra/Serializer/CollectionFiltersNormalizer.php [hunk 3]
```diff
@@ -58,13 +60,19 @@ public function hasCacheableSupportsMethod(): bool
      */
     public function normalize(mixed $object, string $format = null, array $context = []): array|string|int|float|bool|\ArrayObject|null
     {
-        $data = $this->collectionNormalizer->normalize($object, $format, $context);
-        if (!\is_array($data)) {
-            throw new UnexpectedValueException('Expected data to be an array');
+        $preserveEmpty = ($context[Serializer::EMPTY_ARRAY_AS_OBJECT] ?? false) || ($context[AbstractObjectNormalizer::PRESERVE_EMPTY_OBJECTS] ?? false);
```
[the remark this answers is not in the document]
Comment (soyuka) [reply src/Hydra/Serializer/CollectionFiltersNormalizer.php]: mb you're right

File: src/Metadata/Resource/ResourceMetadataCollection.php (hunk 2 above)
Comment (alanpoulain) [reply src/Metadata/Resource/ResourceMetadataCollection.php:58]: I'm not sure to follow: if we ask a `null` `$operationName` to this `getOperation` method, it means we don't know the operation name. It it is cached, the same operation is always returned, even if the other conditions have changed (`$isGetOperation`, `$forceCollection`...). It seems dangerous to me.

File: src/Metadata/Resource/ResourceMetadataCollection.php [hunk 4]
```diff
@@ -24,6 +24,8 @@
  */
 final class ResourceMetadataCollection extends \ArrayObject
 {
+    private const GRAPHQL_PREFIX = 'graphql_';
+    private const HTTP_PREFIX = 'http_';
```
Comment (dunglas) [inline src/Metadata/Resource/ResourceMetadataCollection.php:28]:
```suggestion
    private const GRAPHQL_PREFIX = 'g_';
    private const HTTP_PREFIX = 'h_';
```

As it's entirely internal, let's save a few bytes.

File: src/Serializer/AbstractCollectionNormalizer.php [hunk 5]
```diff
@@ -82,6 +87,23 @@ public function normalize(mixed $object, string $format = null, array $context =
         return array_merge_recursive($data, $paginationData, $itemsData);
     }
 
+    /**
+     * Normalizes a raw collection (not API resources).
+     */
+    protected function normalizeRawCollection(iterable $object, string $format = null, array $context = []): array|\ArrayObject
+    {
+        if (\is_array($object) && !$object && ($context[Serializer::EMPTY_ARRAY_AS_OBJECT] ?? false)) {
```
Comment (dunglas) [inline src/Serializer/AbstractCollectionNormalizer.php:95]: The call to `is_array` is probably the slower, the put it at the end.

File: src/Serializer/AbstractItemNormalizer.php [hunk 6]
```diff
@@ -511,13 +513,22 @@ protected function getFactoryOptions(array $context): array
...
-            $operation = $context['operation'] ?? $this->resourceMetadataCollectionFactory->create($resourceClass)->getOperation($context['operation_name'] ?? null);
-            $options['normalization_groups'] = $operation->getNormalizationContext()['groups'] ?? null;
-            $options['denormalization_groups'] = $operation->getDenormalizationContext()['groups'] ?? null;
-            $options['operation_name'] = $operation->getName();
+        // This is a hot spot
+        if (isset($context['resource_class'])) {
+            $operationCacheKey = $context['resource_class'].($context['operation_name'] ?? '');
+            $operation = $context['operation'] ?? $this->localOperationCache[$operationCacheKey] ?? null;
+
+            if (!isset($this->localOperationCache[$operationCacheKey]) && !$operation && $this->resourceClassResolver->isResourceClass($context['resource_class']) && $this->resourceMetadataCollectionFactory) {
```
Comment (dunglas) [inline src/Serializer/AbstractItemNormalizer.php:521]: Same here, `$this->resourceMetadataCollectionFactory()` is the slowest so should be called last.

File: src/Metadata/Resource/ResourceMetadataCollection.php (hunk 2 above)
Comment (soyuka) [reply src/Metadata/Resource/ResourceMetadataCollection.php:58]: Let's fix this in another patch then.

**Review by alanpoulain — APPROVED** [review]

**Pull request merged**